Add hit testing support for multiline

canavandl opened the pull request:
Adds hit test support for MultiLine glyph

- [x] issues: fixes #3110

canavandl commented on bokehjs/src/coffee/models/glyphs/multi_line.coffee:
```diff
@@ -41,6 +42,79 @@ class MultiLineView extends Glyph.View
+        Object with the following keys:
+          * 0d (bool): whether the point hits the glyph or not
+          * 1d (array(int)): array with the indices hit by the point
+    ###
```
need to update docstring

canavandl commented on bokehjs/src/coffee/models/glyphs/multi_line.coffee:
```diff
@@ -41,6 +42,79 @@ class MultiLineView extends Glyph.View
+    for i in [0...values.length]
+      for j in [0...values[i].length-1]
+        if values[i][j] <= val <= values[i+1]
+          result['2d'].indices.push(i)
```
should also set `result['2d'].point_indices` using nearest points

canavandl commented:
From the example: 

<img width="976" alt="screen shot 2016-07-28 at 2 58 26 pm" src="https://cloud.githubusercontent.com/assets/2198981/17227396/c8ef9952-54d3-11e6-993f-4f49a33c52ce.png">

(There's also a configured HoverTool)

canavandl commented:
I added an example to examples/plotting/file. Currently we don't test hit-testing very well - we test the tap and hover tools, but not the hit-test boundaries of glyphs - so I'll open an issue to begin discussion about that.

canavandl commented on bokehjs/src/coffee/common/hittest.coffee:
```diff
@@ -35,7 +35,13 @@ class HitTestResult
+      # object containing glyph_index: [point_index] of the nearest
+      # point to the crossed segment
+      point_indices: {}
+    }
```
I think this is the only "controversial" part. There has to be an "indices"/array pair to support selections, but it'd be useful to know which points inside a line are selected. My solution was the `point_indices` object which contains the multi_line index and an array of line points.

I'm open to feedback on changing the implementation and/or naming.

bryevdv replied:
we should talk this over. My original intention was that for things like `MultiLine`:

* 1d indices would say "which line" 
* 2d indices would say "which segment" on the lines

Also instead of "nearest" I think for consistency we chose the first point of the segment. But thinking about it now, maybe this would work:
```
@['2d'] = {
    3 : [5],
    6 : [8, 20, 55]
}
```
that would say: lines 3 and 8 were hit, and the values give the segment points. I think we'd still want to fill in the "1d" attr as well (with 3 and 8)

canavandl replied:
>
1d indices would say "which line"
2d indices would say "which segment" on the lines

The hittesting module didn't seem written with that pattern in mind (it was performing lookups for `results['2d'].indices`, but I see thats how hit-testing for patches was implemented. So I changed the multiline hit-testing to match and cleaned up some old code.

I like having the '2d' value be an object better anyway.

canavandl commented:
>
1d indices would say "which line"
2d indices would say "which segment" on the lines

The hittesting module didn't seem written with that pattern in mind (it was performing lookups for `results['2d'].indices`, but I see thats how hit-testing for patches was implemented. So I changed the multiline hit-testing to match and cleaned up some old code.

I like having the '2d' value be an object better anyway.

canavandl commented on bokeh/models/sources.py:
```diff
@@ -22,7 +22,7 @@ class DataSource(Model):
         '0d': {'glyph': None, 'indices': []},
         '1d': {'indices': []},
-        '2d': {'indices': []}
+        '2d': {}
```
'2d' 's value is only an object now

canavandl commented on bokehjs/src/coffee/models/renderers/glyph_renderer.coffee:
```diff
@@ -139,8 +139,6 @@ class GlyphRendererView extends Renderer.View
         selected = indices
       else if selected['1d'].indices.length > 0
         selected = selected['1d'].indices
-      else if selected['2d'].indices.length > 0
```
example of old code that was looking for `indices` attr on `'2d'` element

jbednar commented:
Looks great to me!

bryevdv commented:
I've added `MIGRATION` tag since this could potentially affect people (I doubt many, but still). This LGTM but would appreciate a few more looks @birdsarah @fpliger @mattpap please comment if you can.

birdsarah commented:
I will take a look today or over weekend.
